File: runme/mcgen/local/datan/solsrc/s1lsq.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Table of constant values */

static integer c__9 = 9;
static double c_d1 = 1.;

main(void)
{
    /* Local variables */
    static integer i, k, iexp, lcapt, ltx, lty, nws, nexp, nh, npl, nr;
    static double a[27] /* was [9][3] */, cx[9] /* was [3][3] */, delh, fact,
                  deltay[9], hist[100], p, r, t[9], x[3], y[9], sigmaf, sigmag,
                  scrat[9] /* was [3][3] */, h0, h1, xin[3], xpl[2], ypl[2];
    static char capt[75], tx[75], ty[75];
    static logical ok;
    /* System generated locals */
    integer i__1, i__2, i__3;
/* identify program to user */
    printf("%s\n", "Program S1LSQ generates data");
    printf("%s\n", "corresponding to 2nd degree polynomial");
    printf("%s\n\n", "and fits 1st degree polynomial.");
/* ask for input */
    printf("%s\n", "Enter number NEXP of experiments (>>1):");
    printf("%s", "> ");
    scanf("%li", &nexp);
    printf("%s\n", "Enter X1 for points to be generated:");
    printf("%s", "> ");
    scanf("%lf", &xin[0]);
    printf("%s\n", "Enter X2 for points to be generated:");
    printf("%s", "> ");
    scanf("%lf", &xin[1]);
    printf("%s\n", "Enter X3 for points to be generated:");
    printf("%s", "> ");
    scanf("%lf", &xin[2]);
    printf("%s\n", "Enter SIGMA for points to be generated (>0.):");
    printf("%s", "> ");
    scanf("%lf", &sigmag);
    printf("%s\n", "Enter DELTAY to be used in fit (>0.):");
    printf("%s", "> ");
    scanf("%lf", &sigmaf);
/* ask for number of workstation */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    (void)getchar();
/* initialize histogram and prepare graphics */
    h1 = 1.;
    memset(tx, ' ', 75);
    memset(ty, ' ', 75);
    strncpy(tx, "P(&c@^2#)", 9);
    strncpy(ty, "N(P)", 4);
    ltx = 9;
    lty = 4;
    sprintf(capt, "Input: x_1#=%6.2lf, x_2#=%6.2lf, x_3#=%6.2lf,"
            " &s=%6.2lf, &D@y=%6.2lf", xin[0], xin[1], xin[2], sigmag, sigmaf);
    lcapt = strlen(capt);
    h0 = 0.;
    delh = h1 * .01;
    nh = 100;
/* curve of constant probability */
    fact = (h1 - h0) * (double) nexp / (double) nh;
    xpl[0] = h0;
    xpl[1] = h1;
    ypl[0] = fact;
    ypl[1] = ypl[0];
    npl = 2;
    smhsin_(hist, &h0, &delh, &nh);
/* loop over all simulated experiments */
    i__1 = nexp;
    for (iexp = 1; iexp <= i__1; ++iexp) {
/* generate data points corresponding to polynomial */
        nr = 3;
        rnstnr_(deltay, &c__9);
        for (i = 1; i <= 9; ++i) {
            t[i - 1] = (double) (i - 1) * 1. - 4.;
            y[i - 1] = xin[0];
            if (nr > 1) {
                i__2 = nr;
                for (k = 2; k <= i__2; ++k) {
                    i__3 = k - 1;
                    y[i - 1] += xin[k - 1] * pow(t[i - 1], (double)i__3);
                }
            }
            y[i - 1] += deltay[i - 1] * sigmag;
            deltay[i - 1] = sigmaf;
        }
/* perform polynomial fit */
        nr = 2;
        lsqpol_(t, y, deltay, &c__9, &nr, x, cx, &r, a, scrat, &ok);
/* compute chi-squared probability and enter it in histogram */
        i__2 = 9 - nr;
        p = 1. - scchi2_(&r, &i__2);
        smhsfl_(hist, &h0, &delh, &nh, &p, &c_d1);
    }
/* graphical output */
    grhscv_(xpl, ypl, &npl, hist, &h0, &delh, &nh, tx, &ltx, ty, &lty, capt,
           &lcapt, &nws);
    return 0;
} /* main */

File: runme/mcgen/local/datan/solsrc/s1lsq.f
      PROGRAM S1LSQ
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(MAXNR=3,MAXN=9,N=9,T0=-4.D0,DELTAT=1.D0)
      DIMENSION T(MAXN),DELTAY(MAXN),CX(MAXNR,MAXNR),A(MAXN,MAXNR)
      DIMENSION Y(MAXN),X(MAXNR),SCRAT(MAXNR,MAXNR),HIST(100)
      DIMENSION XIN(MAXNR)
      DIMENSION XPL(2),YPL(2)
      CHARACTER*75 TX,TY,CAPT
      LOGICAL OK
C identify program to user
      WRITE(*,*)' Program S1LSQ generates data'
      WRITE(*,*)' corresponding to 2nd degree polynomial'
      WRITE(*,*)' and fits 1st degree polynomial'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)' Enter number NEXP of experiments (>>1)'
      WRITE(*,*)' >'
      READ(*,*)NEXP
      WRITE(*,*)' Enter X1 for points to be generated'
      WRITE(*,*)' >'
      READ(*,*)XIN(1)
      WRITE(*,*)' Enter X2 for points to be generated'
      WRITE(*,*)' >'
      READ(*,*)XIN(2)
      WRITE(*,*)' Enter X3 for points to be generated'
      WRITE(*,*)' >'
      READ(*,*)XIN(3)
      WRITE(*,*)' Enter SIGMA for points to be generated (>0.)'
      WRITE(*,*)' >'
      READ(*,*)SIGMAG
      WRITE(*,*)' Enter DELTAY to be used in fit (>0.)'
      WRITE(*,*)' >'
      READ(*,*)SIGMAF
C ask for number of workstation
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
C initialize histogram and prepare graphics
      H1=1.D0
      TX='P(&c@^2#)'
      TY='N(P)'
      LTX=9
      LTY=4
      CAPT(1:45)='Input: x_1#=******, x_2#=******, x_3#=******,'
      CAPT(46:68)=' &s=******, &D@y=******'
      WRITE(CAPT(13:18),'(F6.2)')XIN(1)
      WRITE(CAPT(26:31),'(F6.2)')XIN(2)
      WRITE(CAPT(39:44),'(F6.2)')XIN(3)
      WRITE(CAPT(50:55),'(F6.2)')SIGMAG
      WRITE(CAPT(63:68),'(F6.2)')SIGMAF
      LCAPT=68
      H0=0.D0
      DELH=0.01D0*H1
      NH=100
C curve of constant probability
      FACT=(H1-H0)*DBLE(NEXP)/DBLE(NH)
      XPL(1)=H0
      XPL(2)=H1
      YPL(1)=FACT
      YPL(2)=YPL(1)
      NPL=2
      CALL SMHSIN(HIST,H0,DELH,NH)
C loop over all simulated experiments
      DO 100 IEXP=1,NEXP
C generate data points corresponding to polynomial
        NR=3
        CALL RNSTNR(DELTAY,N)
        DO 20 I=1,N
          T(I)=T0+DBLE(I-1)*DELTAT
          Y(I)=XIN(1)
          IF(NR.GT.1) THEN
            DO 10 K=2,NR
              Y(I)=Y(I)+XIN(K)*T(I)**(K-1)
10          CONTINUE
          END IF
          Y(I)=Y(I)+DELTAY(I)*SIGMAG
          DELTAY(I)=SIGMAF
20      CONTINUE
C perform polynomial fit
        NR=2
        CALL LSQPOL(T,Y,DELTAY,N,NR,X,CX,R,A,SCRAT,OK)
C compute chi-squared probability and enter it in histogram
        P=1.D0-SCCHI2(R,N-NR)
        CALL SMHSFL(HIST,H0,DELH,NH,P,1.D0)
100   CONTINUE
C graphical output
      CALL GRHSCV(XPL,YPL,NPL,HIST,H0,DELH,NH,TX,LTX,TY,LTY,CAPT,
     +LCAPT,NWS)
      END
